File: api/app/services/report_service.py
"""
Report service for analytics and statistics.
"""
from datetime import date
from decimal import Decimal

from sqlalchemy import Integer, func
from sqlalchemy.orm import Session

from app.models.apartment import Apartment, Room, RoomStatus
from app.models.bill import Bill, BillStatus
from app.models.lease import Lease
from app.models.tenant import Tenant
from app.services.base import BaseService


class ReportService(BaseService):
    """Service for reports and analytics."""

    def __init__(self, db: Session):
        super().__init__(db)

    def get_overview(self, org_id: int) -> dict:
        """Get dashboard overview statistics."""
        # Count apartments
        total_apartments = (
            self.db.query(Apartment)
            .filter(Apartment.organization_id == org_id)
            .count()
        )

        # Count rooms
        total_rooms = (
            self.db.query(Room)
            .join(Apartment)
            .filter(Apartment.organization_id == org_id)
            .count()
        )

        # Count occupied rooms
        occupied_rooms = (
            self.db.query(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Room.status == RoomStatus.OCCUPIED,
            )
            .count()
        )

        # Available rooms
        available_rooms = total_rooms - occupied_rooms

        # Occupancy rate
        occupancy_rate = round((occupied_rooms / total_rooms) * 100, 1) if total_rooms > 0 else 0

        # Active leases
        active_leases = (
            self.db.query(Lease)
            .join(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Lease.is_active.is_(True),
            )
            .count()
        )

        # Total tenants
        total_tenants = (
            self.db.query(Tenant)
            .filter(Tenant.organization_id == org_id)
            .count()
        )

        # Monthly revenue (current month)
        today = date.today()
        monthly_revenue = (
            self.db.query(func.sum(Bill.total_amount))
            .join(Lease)
            .join(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Bill.bill_year == today.year,
                Bill.bill_month == today.month,
            )
            .scalar() or Decimal(0)
        )

        # Pending bills
        pending_bills = (
            self.db.query(Bill)
            .join(Lease)
            .join(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Bill.status == BillStatus.PENDING,
            )
            .count()
        )

        # Overdue bills
        overdue_bills = (
            self.db.query(Bill)
            .join(Lease)
            .join(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Bill.status != BillStatus.PAID,
                Bill.due_date < today,
            )
            .count()
        )

        return {
            "total_apartments": total_apartments,
            "total_rooms": total_rooms,
            "occupied_rooms": occupied_rooms,
            "available_rooms": available_rooms,
            "occupancy_rate": occupancy_rate,
            "active_leases": active_leases,
            "total_tenants": total_tenants,
            "monthly_revenue": float(monthly_revenue),
            "pending_bills": pending_bills,
            "overdue_bills": overdue_bills,
        }

    def get_income_report(
        self,
        org_id: int,
        year: int,
        start_month: int | None = None,
        end_month: int | None = None,
    ) -> list:
        """Get income report by month.

        Returns a list of monthly income reports with breakdown by category.
        """
        query = (
            self.db.query(
                Bill.bill_month,
                func.sum(Bill.total_amount).label("total_amount"),
                func.sum(Bill.paid_amount).label("collected_amount"),
                func.sum(Bill.rent_amount).label("total_rent"),
                func.sum(Bill.water_amount).label("total_water"),
                func.sum(Bill.electricity_amount).label("total_electricity"),
                func.sum(Bill.other_amount).label("total_other"),
            )
            .join(Lease)
            .join(Room)
            .join(Apartment)
            .filter(
                Apartment.organization_id == org_id,
                Bill.bill_year == year,
            )
        )

        if start_month:
            query = query.filter(Bill.bill_month >= start_month)
        if end_month:
            query = query.filter(Bill.bill_month <= end_month)

        results = query.group_by(Bill.bill_month).all()

        monthly_data = []
        for row in results:
            total_amount = float(row.total_amount or 0)
            collected_amount = float(row.collected_amount or 0)
            monthly_data.append({
                "period": f"{row.bill_month}月",
                "total_amount": total_amount,
                "collected_amount": collected_amount,
                "total_rent": float(row.total_rent or 0),
                "total_water": float(row.total_water or 0),
                "total_electricity": float(row.total_electricity or 0),
                "total_other": float(row.total_other or 0),
                "collection_rate": round((collected_amount / total_amount) * 100, 1) if total_amount > 0 else 0,
            })

        return monthly_data

    def get_occupancy_report(self, org_id: int, year: int) -> list:
        """Get occupancy report by month.

        Calculate monthly occupancy rate based on active leases.
        """
        # Get total rooms count
        total_rooms = (
            self.db.query(Room)
            .join(Apartment)
            .filter(Apartment.organization_id == org_id)
            .count()
        )

        if total_rooms == 0:
            return []

        monthly_data = []
        for month in range(1, 13):
            # A room is occupied in a month if there's an active lease
            # that covers that month
            month_start = date(year, month, 1)
            if month == 12:
                month_end = date(year + 1, 1, 1)
            else:
                month_end = date(year, month + 1, 1)

            # Active lease: start_date <= month_end AND (end_date IS NULL OR end_date >= month_start)
            occupied_rooms = (
                self.db.query(func.count(func.distinct(Room.id)))
                .join(Lease)
                .join(Apartment)
                .filter(
                    Apartment.organization_id == org_id,
                    Lease.start_date < month_end,
                    func.coalesce(Lease.end_date, date(2999, 12, 31)) >= month_start,
                )
                .scalar() or 0
            )

            occupancy_rate = round((occupied_rooms / total_rooms) * 100, 1) if total_rooms > 0 else 0

            monthly_data.append({
                "period": f"{month}月",
                "total_rooms": total_rooms,
                "occupied_rooms": occupied_rooms,
                "vacant_rooms": total_rooms - occupied_rooms,
                "occupancy_rate": occupancy_rate,
            })

        return monthly_data

File: api/src/services/report.service.ts
import type { ReportRepository } from '../repositories/report.repo.js';
import { defaultReportRepo } from '../repositories/report.repo.js';

/**
 * 概览统计响应
 */
export interface OverviewStats {
  total_apartments: number;
  total_rooms: number;
  occupied_rooms: number;
  available_rooms: number;
  total_tenants: number;
  active_leases: number;
  occupancy_rate: number;
  monthly_revenue: number;
  pending_bills: number;
  overdue_bills: number;
  rooms_missing_initial_readings: number;
}

/**
 * 月度收入统计
 */
export interface MonthlyIncome {
  period: string;
  total_rent: number;
  total_water: number;
  total_electricity: number;
  total_other: number;
  total_amount: number;
  collected_amount: number;
  collection_rate: number;
}

/**
 * 月度入住率统计
 */
export interface MonthlyOccupancy {
  period: string;
  total_rooms: number;
  occupied_rooms: number;
  vacant_rooms: number;
  occupancy_rate: number;
}

/**
 * Report Service 接口
 */
export interface ReportService {
  getOverview(orgId: string): Promise<OverviewStats>;
  getIncome(
    orgId: string,
    year: number,
    startMonth?: number,
    endMonth?: number
  ): Promise<MonthlyIncome[]>;
  getOccupancy(orgId: string, year: number): Promise<MonthlyOccupancy[]>;
}

/**
 * 创建 Report Service 实例
 */
export function createReportService(
  getRepo: () => ReportRepository = () => defaultReportRepo
): ReportService {
  return {
    getOverview: async (orgId: string) => {
      const today = new Date();
      const [
        total_apartments,
        total_rooms,
        occupied_rooms,
        total_tenants,
        active_leases,
        monthly_revenue,
        pending_bills,
        overdue_bills,
        rooms_missing_initial_readings,
      ] = await Promise.all([
        getRepo().countApartments(orgId),
        getRepo().countRooms(orgId),
        getRepo().countOccupiedRooms(orgId),
        getRepo().countTenants(orgId),
        getRepo().countActiveLeases(orgId),
        getRepo().sumMonthlyRevenue(orgId, today.getFullYear(), today.getMonth() + 1),
        getRepo().countPendingBills(orgId),
        getRepo().countOverdueBills(orgId, today),
        getRepo().getRoomsMissingInitialReadings(orgId),
      ]);

      const available_rooms = total_rooms - occupied_rooms;
      const occupancy_rate =
        total_rooms > 0 ? Math.round((occupied_rooms / total_rooms) * 1000) / 10 : 0;

      return {
        total_apartments,
        total_rooms,
        occupied_rooms,
        available_rooms,
        total_tenants,
        active_leases,
        occupancy_rate,
        monthly_revenue,
        pending_bills,
        overdue_bills,
        rooms_missing_initial_readings,
      };
    },

    getIncome: async (orgId: string, year: number, startMonth?: number, endMonth?: number) => {
      const roomIds = await getRepo().getRoomIds(orgId);
      const leaseIds = await getRepo().getLeaseIds(roomIds);
      const bills = await getRepo().getBillsByYear(leaseIds, year, startMonth, endMonth);

      const byMonth = new Map<
        number,
        {
          total_rent: number;
          total_water: number;
          total_electricity: number;
          total_other: number;
          total_amount: number;
          collected_amount: number;
        }
      >();

      for (const b of bills) {
        const m = b.bill_month;
        if (!byMonth.has(m)) {
          byMonth.set(m, {
            total_rent: 0,
            total_water: 0,
            total_electricity: 0,
            total_other: 0,
            total_amount: 0,
            collected_amount: 0,
          });
        }
        const row = byMonth.get(m)!;
        row.total_rent += Number(b.rent_amount);
        row.total_water += Number(b.water_amount);
        row.total_electricity += Number(b.electricity_amount);
        row.total_other += Number(b.other_amount);
        row.total_amount += Number(b.total_amount);
        row.collected_amount += Number(b.paid_amount);
      }

      const sortedMonths = Array.from(byMonth.entries()).sort((a, b) => a[0] - b[0]);
      return sortedMonths.map(([month, row]) => ({
        period: `${month}月`,
        total_rent: row.total_rent,
        total_water: row.total_water,
        total_electricity: row.total_electricity,
        total_other: row.total_other,
        total_amount: row.total_amount,
        collected_amount: row.collected_amount,
        collection_rate:
          row.total_amount > 0
            ? Math.round((row.collected_amount / row.total_amount) * 1000) / 10
            : 0,
      }));
    },

    getOccupancy: async (orgId: string, year: number) => {
      const total_rooms = await getRepo().countRoomsTotal(orgId);
      if (total_rooms === 0) {
        return [];
      }

      const roomIds = await getRepo().getRoomIds(orgId);
      const result: MonthlyOccupancy[] = [];

      for (let month = 1; month <= 12; month++) {
        const monthStart = new Date(year, month - 1, 1);
        const monthEndNext = new Date(year, month, 1);

        const occupiedInMonth = await getRepo().countOccupiedRoomsInMonth(
          roomIds,
          monthStart,
          monthEndNext
        );
        const vacant_rooms = total_rooms - occupiedInMonth;
        const occupancy_rate =
          total_rooms > 0 ? Math.round((occupiedInMonth / total_rooms) * 1000) / 10 : 0;

        result.push({
          period: `${month}月`,
          total_rooms,
          occupied_rooms: occupiedInMonth,
          vacant_rooms,
          occupancy_rate,
        });
      }

      return result;
    },
  };
}

/**
 * 默认实例
 */
export const defaultReportService = createReportService();
